src/Storages/StorageS3Distributed.h

```diff
@@ -0,0 +1,157 @@
+#pragma once
+
+#include <Common/config.h>
+
+#if USE_AWS_S3
+
+#include "Client/Connection.h"
+#include "Interpreters/Cluster.h"
+#include "Storages/IStorage.h"
+#include "Storages/StorageS3.h"
+
+#include <memory>
+#include <optional>
+#include "ext/shared_ptr_helper.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+
+class Context;
+
+struct ClientAuthentificationBuilder
+{
+    String access_key_id;
+    String secret_access_key;
+    UInt64 max_connections;
+};
+
+using QueryId = std::string;
+using Task = std::string;
+using Tasks = std::vector<Task>;
+using TasksIterator = Tasks::iterator;
+
+class S3NextTaskResolver
+{
+public:
+    S3NextTaskResolver(QueryId query_id, Tasks && all_tasks)
+        : id(query_id)
+        , tasks(all_tasks)
+        , current(tasks.begin())
+    {}
+
+    std::string next()
+    {
+        auto it = current;
+        ++current;
+        return it == tasks.end() ? "" : *it;
+    }
+
+    std::string getId()
+    {
+        return id;
+    }
+
+private:
+    QueryId id;
```

Comment:
It may be not correct to use query id.
E.g. if we read like `select * from s3dist union all select * from s3dist`.
Query will be the same, and tables will be the same. We need to enumerate storage in query also.

Reply:
I think it is better to generate UUID

Reply:
Seems like Fixed